app/src/main/java/nodomain/freeyourgadget/gadgetbridge/service/devices/hplus/HPlusSupport.java

```diff
@@ -0,0 +1,881 @@
+            } else {
+                message = title;
+                for (int i = message.length(); i < 12; i++)
+                    message += " ";
+            }
+        }
+        message += body;
+
+        int length = message.length() / 17;
+
+        builder.write(ctrlCharacteristic, new byte[]{HPlusConstants.COMMAND_ACTION_INCOMING_SOCIAL, (byte) 255});
+
+        int remaining = 0;
+
+        if (message.length() % 17 > 0)
+            remaining = length + 1;
+        else
+            remaining = length;
+
+        msg[1] = (byte) remaining;
+        int message_index = 0;
+        int i = 3;
+
+        for (int j = 0; j < message.length(); j++) {
+            msg[i++] = (byte) message.charAt(j);
+
+            if (i == msg.length) {
+                message_index++;
+                msg[2] = (byte) message_index;
+                builder.write(ctrlCharacteristic, msg);
+
+                msg = msg.clone();
+                for (i = 3; i < msg.length; i++)
+                    msg[i] = 32;
+
+                if (message_index < remaining)
+                    i = 3;
+                else
+                    break;
+            }
+        }
+
+        msg[2] = (byte) remaining;
+
+        builder.write(ctrlCharacteristic, msg);
+        builder.queue(getQueue());
+    }
+
+    public boolean isExpectedDevice(BluetoothDevice device) {
+        return true;
+    }
+
+    public void close() {
+    }
+
+    @Override
+    public boolean onCharacteristicChanged(BluetoothGatt gatt,
+                                           BluetoothGattCharacteristic characteristic) {
+        if (super.onCharacteristicChanged(gatt, characteristic)) {
+            return true;
+        }
+
+        UUID characteristicUUID = characteristic.getUuid();
+        byte[] data = characteristic.getValue();
+        if (data.length == 0)
+            return true;
+
+        switch (data[0]) {
+            case HPlusConstants.DATA_STATS:
+                return processDataStats(data);
+            case HPlusConstants.DATA_SLEEP:
+                return processSleepStats(data);
+            default:
+                LOG.info("Unhandled characteristic changed: " + characteristicUUID);
+
+        }
+        return false;
+    }
+
+    private boolean processSleepStats(byte[] data) {
+        LOG.debug("Process Sleep Stats");
+
+        if (data.length < 19) {
+            LOG.error("Invalid Sleep Message Length " + data.length);
+            return false;
+        }
+        HPlusSleepRecord record = new HPlusSleepRecord(data);
+
+        try (DBHandler handler = GBApplication.acquireDB()) {
+            DaoSession session = handler.getDaoSession();
+
+            Device device = DBHelper.getDevice(getDevice(), session);
+            User user = DBHelper.getUser(session);
+            int ts = (int) (System.currentTimeMillis() / 1000);
+            HPlusSampleProvider provider = new HPlusSampleProvider(gbDevice, session);
+
+            //TODO: Store Sample. How?
```

Comment:
You can store sleep information the same way as in the method below. I'd buffer the samples locally before committing them to the database. This would allow combining activity samples with sleep, but this depends on how the device sends the data.